### PR TITLE
Enable-WriteCore-for-Azure.AI.Translation.Document

### PR DESCRIPTION
**Pull request by welovej**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs [hunk 1]
```diff
@@ -35,8 +36,9 @@ public DocumentTranslateContent(Azure.AI.Translation.Document.MultipartFormFileD
...
         public Azure.AI.Translation.Document.MultipartFormFileData MultipartDocument { get { throw null; } }
         public System.Collections.Generic.IList<Azure.AI.Translation.Document.MultipartFormFileData> MultipartGlossary { get { throw null; } }
-        void global::System.ClientModel.Primitives.IJsonModel<Azure.AI.Translation.Document.DocumentTranslateContent>.Write(System.Text.Json.Utf8JsonWriter writer, System.ClientModel.Primitives.ModelReaderWriterOptions options) { }
+        protected virtual void JsonModelWriteCore(System.Text.Json.Utf8JsonWriter writer, System.ClientModel.Primitives.ModelReaderWriterOptions options) { }
         Azure.AI.Translation.Document.DocumentTranslateContent System.ClientModel.Primitives.IJsonModel<Azure.AI.Translation.Document.DocumentTranslateContent>.Create(ref System.Text.Json.Utf8JsonReader reader, System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        void System.ClientModel.Primitives.IJsonModel<Azure.AI.Translation.Document.DocumentTranslateContent>.Write(System.Text.Json.Utf8JsonWriter writer, System.ClientModel.Primitives.ModelReaderWriterOptions options) { }
```
Comment (pshao25) [inline sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:41]: Is this actually no change right? @ArcturusZhang

Comment (ArcturusZhang) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:41]: The API does not change here.

File: sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslateContent.Serialization.cs [hunk 2]
```diff
@@ -27,14 +27,15 @@ internal virtual MultipartFormDataRequestContent ToMultipartRequestContent()
...
+        protected virtual void JsonModelWriteCore(Utf8JsonWriter writer, ModelReaderWriterOptions options)
         {
             var format = options.Format == "W" ? ((IPersistableModel<DocumentTranslateContent>)this).GetFormatFromOptions(options) : options.Format;
             if (format != "J")
             {
                 throw new FormatException($"The model {nameof(DocumentTranslateContent)} does not support writing '{format}' format.");
             }
-
             writer.WriteStartObject();
             writer.WritePropertyName("document"u8);
```
Comment (ArcturusZhang) [inline sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslateContent.Serialization.cs:40]: this change is not correct.
Our design is:
The implementation of the interface (Write method) will call the WriteCore method, wrapping it by `WriteStartObject` and `WriteEndObject`, and the implementation of the WriteCore method should NOT contain the `WriteStartObject` and `WriteEndObject`.
Therefore this part needs to change to remove the `WriteStartObject` and `WriteEndObject`.

File: sdk/translation/Azure.AI.Translation.Document/src/Generated/DocumentTranslateContent.Serialization.cs [hunk 3]
```diff
@@ -17,6 +17,13 @@ public partial class DocumentTranslateContent : IUtf8JsonSerializable, IJsonMode
     {
         void IUtf8JsonSerializable.Write(Utf8JsonWriter writer) => ((IJsonModel<DocumentTranslateContent>)this).Write(writer, ModelSerializationExtensions.WireOptions);
 
+        void IJsonModel<DocumentTranslateContent>.Write(Utf8JsonWriter writer, ModelReaderWriterOptions options)
+        {
+            writer.WriteStartObject();
+            JsonModelWriteCore(writer, options);
+            writer.WriteEndObject();
+        }
```
Comment (ArcturusZhang) [inline sdk/translation/Azure.AI.Translation.Document/src/Generated/DocumentTranslateContent.Serialization.cs:25]: As I said in [this comment](https://github.com/Azure/azure-sdk-for-net/pull/46895/files#r1825357024), if you follow the invocation starting here, you will find that the `start object` and `end object` are written twice, which is incorrect.

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]
Since I notice there is an issue in the serialization of `DocumentTranslateContent` class, could we introduce a test case?
This test case does not have to be pushed into this PR, you could have this test and verify its serialized JSON is expected and valid

File: sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslateContent.Serialization.cs (hunk 2 above)
Comment (welovej) [reply sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslateContent.Serialization.cs]: Fixed

File: sdk/translation/Azure.AI.Translation.Document/src/Generated/DocumentTranslateContent.Serialization.cs (hunk 3 above)
Comment (welovej) [reply sdk/translation/Azure.AI.Translation.Document/src/Generated/DocumentTranslateContent.Serialization.cs:25]: Fixed

File: sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs (hunk 1 above)
Comment (welovej) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:41]: Fixed

File: sdk/translation/Azure.AI.Translation.Document/tests/ContainerDocumentTranslationClientTests.cs [hunk 4]
```diff
@@ -47,5 +48,23 @@ public async Task StartSynchronousDocumentTranslation()
             Assert.IsNotNull(responseString);
             Assert.AreNotEqual(requestString, responseString);
         }
+
+        [Test]
+        public void DocumentTranslationSerializationTest()
```
Comment (pshao25) [inline sdk/translation/Azure.AI.Translation.Document/tests/ContainerDocumentTranslationClientTests.cs:53]: Can we add some comments it is for serialization sanity test, not for functional test

**Review by pshao25 — APPROVED** [review]

File: sdk/translation/Azure.AI.Translation.Document/tests/ContainerDocumentTranslationClientTests.cs (hunk 4 above)
Comment (welovej) [reply sdk/translation/Azure.AI.Translation.Document/tests/ContainerDocumentTranslationClientTests.cs:54]: Fixed

File: sdk/translation/Azure.AI.Translation.Document/tests/ContainerDocumentTranslationClientTests.cs [hunk 5]
```diff
@@ -47,5 +48,24 @@ public async Task StartSynchronousDocumentTranslation()
...
+
+            filePath = Path.Combine("TestData", "test-glossary-2.csv");
+            using Stream glossaryStream = File.OpenRead(filePath);
+            var sourceGlossaries = new List<MultipartFormFileData>()
+            {
+                new(Path.GetFileName(filePath), glossaryStream, "text/csv")
+            };
+            DocumentTranslateContent content = new DocumentTranslateContent(sourceDocument, sourceGlossaries);
+            BinaryData result = ModelReaderWriter.Write<DocumentTranslateContent>(content);
+            Assert.IsNotNull(result);
```
Comment (ArcturusZhang) [inline sdk/translation/Azure.AI.Translation.Document/tests/ContainerDocumentTranslationClientTests.cs:68]: so actually this `ModelReaderWriter.Write<T>` method never returns a null - could we check what is inside the content?

I kind of do not like that we have this test. In my original comment, I am suggesting that we just have a local test verifying everything is expected and we are good to go.

**Review by ArcturusZhang — APPROVED** [review]

**Pull request merged**